File: web/frontend/chromadb.js
const { createApp, ref, computed, onMounted } = Vue;
const { ElMessage, ElMessageBox } = ElementPlus;

const API_BASE_URL = window.location.origin;

const app = createApp({
    setup() {
        const collections = ref([]);
        const selectedCollection = ref('');
        const queryText = ref('');
        const nResults = ref(10);
        const results = ref([]);
        const loading = ref(false);
        const querying = ref(false);
        const loadingSample = ref(false);
        const showingSample = ref(false);
        const currentUsername = ref(localStorage.getItem('username') || '用户');

        const currentCollectionInfo = computed(() => {
            if (!selectedCollection.value) return null;
            return collections.value.find(c => c.name === selectedCollection.value);
        });

        // 获取认证头
        const getAuthHeaders = () => {
            const token = localStorage.getItem('access_token');
            if (!token) {
                return {};
            }
            return {
                'Authorization': `Bearer ${token}`,
                'Content-Type': 'application/json'
            };
        };

        // 检查认证状态
        const checkAuth = () => {
            const token = localStorage.getItem('access_token');
            if (!token) {
                window.location.href = '/login';
                return false;
            }
            return true;
        };

        // 处理认证错误
        const handleAuthError = (error) => {
            if (error.message.includes('401') || error.message.includes('Unauthorized')) {
                ElMessage.error('登录已过期，请重新登录');
                localStorage.removeItem('access_token');
                localStorage.removeItem('username');
                setTimeout(() => {
                    window.location.href = '/login';
                }, 1000);
                return true;
            }
            return false;
        };

        // 检查配置
        const checkConfig = async () => {
            if (!checkAuth()) return;

            try {
                const response = await fetch(`${API_BASE_URL}/api/chromadb/config`, {
                    headers: getAuthHeaders()
                });

                if (!response.ok) {
                    throw new Error(`HTTP ${response.status}`);
                }

                const data = await response.json();
                
                let message = `配置文件: ${data.config_file}\n`;
                if (data.working_directory) {
                    message += `工作目录: ${data.working_directory}\n\n`;
                }
                
                if (data.error) {
                    message += `错误: ${data.error}`;
                } else {
                    if (data.chromadb) {
                        message += `【ChromaDB 配置】\n`;
                        message += `连接类型: ${data.chromadb.connection_type}\n`;
                        if (data.chromadb.connection_type === 'remote') {
                            message += `服务器: ${data.chromadb.host}:${data.chromadb.port}\n`;
                            message += `SSL: ${data.chromadb.ssl_enabled}\n`;
                        } else {
                            message += `路径: ${data.chromadb.path}\n`;
                        }
                    }
                    
                    if (data.embedding_api) {
                        message += `\n【Embedding API】\n`;
                        if (data.embedding_api.configured) {
                            message += `端点: ${data.embedding_api.endpoint}\n`;
                            message += `模型: ${data.embedding_api.model}\n`;
                            message += `状态: ✅ 已配置`;
                        } else {
                            message += `状态: ❌ 未配置`;
                        }
                    }
                }
                
                ElMessageBox.alert(message, 'ChromaDB 配置信息', {
                    confirmButtonText: '确定'
                });
            } catch (error) {
                console.error('检查配置失败:', error);
                if (!handleAuthError(error)) {
                    ElMessage.error('检查配置失败: ' + error.message);
                }
            }
        };

        // 加载集合列表
        const loadCollections = async () => {
            if (!checkAuth()) return;

            loading.value = true;
            try {
                const response = await fetch(`${API_BASE_URL}/api/chromadb/collections`, {
                    headers: getAuthHeaders()
                });

                if (!response.ok) {
                    const errorData = await response.json().catch(() => ({}));
                    throw new Error(errorData.detail || `HTTP ${response.status}`);
                }

                const data = await response.json();
                collections.value = data.collections || [];
                
                if (collections.value.length > 0) {
                    ElMessage.success(`加载了 ${collections.value.length} 个集合`);
                } else {
                    ElMessage.warning('未找到任何集合');
                }
            } catch (error) {
                console.error('加载集合失败:', error);
                if (!handleAuthError(error)) {
                    ElMessage.error('加载集合失败: ' + error.message);
                }
            } finally {
                loading.value = false;
            }
        };

        // 集合变化时
        const onCollectionChange = () => {
            results.value = [];
            showingSample.value = false;
            queryText.value = '';
        };

        // 执行查询
        const executeQuery = async () => {
            if (!selectedCollection.value || !queryText.value) {
                ElMessage.warning('请选择集合并输入查询关键词');
                return;
            }

            querying.value = true;
            showingSample.value = false;
            try {
                const response = await fetch(`${API_BASE_URL}/api/chromadb/query`, {
                    method: 'POST',
                    headers: getAuthHeaders(),
                    body: JSON.stringify({
                        collection_name: selectedCollection.value,
                        query_text: queryText.value,
                        n_results: nResults.value
                    })
                });

                if (!response.ok) {
                    const errorData = await response.json();
                    throw new Error(errorData.detail || `HTTP ${response.status}`);
                }

                const data = await response.json();
                results.value = data.results || [];
                
                ElMessage.success(`找到 ${results.value.length} 条结果`);
            } catch (error) {
                console.error('查询失败:', error);
                if (!handleAuthError(error)) {
                    ElMessage.error('查询失败: ' + error.message);
                }
            } finally {
                querying.value = false;
            }
        };

        // 加载示例数据
        const loadSample = async () => {
            if (!selectedCollection.value) {
                ElMessage.warning('请先选择集合');
                return;
            }

            loadingSample.value = true;
            showingSample.value = true;
            try {
                const response = await fetch(
                    `${API_BASE_URL}/api/chromadb/collection/${selectedCollection.value}/sample?limit=5`,
                    {
                        headers: getAuthHeaders()
                    }
                );

                if (!response.ok) {
                    const errorData = await response.json();
                    throw new Error(errorData.detail || `HTTP ${response.status}`);
                }

                const data = await response.json();
                results.value = data.samples || [];
                
                ElMessage.success(`加载了 ${results.value.length} 条示例数据`);
            } catch (error) {
                console.error('加载示例失败:', error);
                if (!handleAuthError(error)) {
                    ElMessage.error('加载示例失败: ' + error.message);
                }
            } finally {
                loadingSample.value = false;
            }
        };

        // 返回历史记录页面
        const goBack = () => {
            window.location.href = '/';
        };

        // 登出
        const logout = async () => {
            try {
                await ElMessageBox.confirm('确定要退出登录吗？', '确认退出', {
                    type: 'warning'
                });

                const token = localStorage.getItem('access_token');
                if (token) {
                    try {
                        await fetch(`${API_BASE_URL}/api/auth/logout`, {
                            method: 'POST',
                            headers: getAuthHeaders()
                        });
                    } catch (error) {
                        console.error('登出请求失败:', error);
                    }
                }

                localStorage.removeItem('access_token');
                localStorage.removeItem('username');
                ElMessage.success('已退出登录');

                setTimeout(() => {
                    window.location.href = '/login';
                }, 500);
            } catch (error) {
                // 用户取消
            }
        };

        onMounted(() => {
            if (!checkAuth()) return;
            loadCollections();
        });

        return {
            collections,
            selectedCollection,
            queryText,
            nResults,
            results,
            loading,
            querying,
            loadingSample,
            showingSample,
            currentCollectionInfo,
            currentUsername,
            checkConfig,
            loadCollections,
            onCollectionChange,
            executeQuery,
            loadSample,
            goBack,
            logout
        };
    }
});

app.use(ElementPlus).mount('#app');

File: web/backend/api/endpoints/chromadb.py
"""
ChromaDB 查询 API 端点
"""

from fastapi import APIRouter, HTTPException
from typing import Optional, List
import chromadb
from chromadb.config import Settings
from pydantic import BaseModel
import configparser
import os
import requests

router = APIRouter(prefix="/api/chromadb", tags=["chromadb"])


class QueryRequest(BaseModel):
    collection_name: str
    query_text: str
    n_results: int = 10


class QueryResult(BaseModel):
    id: str
    document: str
    metadata: dict
    distance: float


def get_config():
    """读取配置文件"""
    config = configparser.ConfigParser()
    current_dir = os.path.dirname(os.path.abspath(__file__))
    
    # 尝试多个可能的配置文件路径
    config_paths = [
        'config.ini',  # 当前工作目录
        os.path.join(os.getcwd(), 'config.ini'),  # 明确的工作目录
        os.path.join(current_dir, '..', '..', 'config.ini'),  # web/backend/config.ini
        os.path.join(current_dir, '..', '..', '..', 'config.ini'),  # 项目根目录
        os.path.join(current_dir, '..', '..', '..', '..', 'config.ini'),  # 再上一级
    ]
    
    for config_path in config_paths:
        abs_path = os.path.abspath(config_path)
        if os.path.exists(abs_path):
            try:
                config.read(abs_path, encoding='utf-8')
                print(f"✓ 成功加载配置文件: {abs_path}")
                return config, abs_path
            except Exception as e:
                print(f"✗ 读取配置文件失败 {abs_path}: {e}")
                continue
    
    print("✗ 未找到配置文件")
    return config, None


def get_embedding_function():
    """根据配置文件获取 embedding 函数"""
    config, config_path = get_config()
    
    if not config_path:
    
    if 'embedding_api' not in config:
        return None
    
    endpoint = config.get('embedding_api', 'endpoint', fallback=None)
    api_key = config.get('embedding_api', 'api_key', fallback=None)
    model = config.get('embedding_api', 'model', fallback=None)
    
    if not all([endpoint, api_key, model]):
        return None
    
    # 创建自定义 embedding 函数
    class CustomEmbeddingFunction:
        def __init__(self, endpoint, api_key, model):
            self.endpoint = endpoint
            self.api_key = api_key
            self.model = model
        
        def __call__(self, input):
            """生成 embeddings"""
            headers = {
                'Authorization': f'Bearer {self.api_key}',
                'Content-Type': 'application/json'
            }
            
            # 确保 input 是列表
            if isinstance(input, str):
                texts = [input]
            else:
                texts = input
            
            payload = {
                'model': self.model,
                'input': texts,
                'encoding_format': 'float'
            }
            
            try:
                response = requests.post(self.endpoint, headers=headers, json=payload, timeout=30)
                response.raise_for_status()
                data = response.json()
                
                # 提取 embeddings
                embeddings = [item['embedding'] for item in data['data']]
                return embeddings
            except Exception as e:
                raise Exception(f"Embedding API 调用失败: {str(e)}")
    
    return CustomEmbeddingFunction(endpoint, api_key, model)


def get_chromadb_client():
    """根据配置文件获取 ChromaDB 客户端"""
    config, config_path = get_config()
    
    if not config_path or 'chromadb' not in config:
        # 默认使用本地连接
        print("未找到配置文件或配置不完整，使用默认本地连接")
        return chromadb.PersistentClient(path="./data/chromadb")
    
    connection_type = config.get('chromadb', 'connection_type', fallback='local')
    
    if connection_type == 'remote':
        host = config.get('chromadb', 'host', fallback='localhost')
        port = config.getint('chromadb', 'port', fallback=8000)
        ssl_enabled = config.getboolean('chromadb', 'ssl_enabled', fallback=False)
        auth_token = config.get('chromadb', 'auth_token', fallback=None)
        
        # 构建远程连接URL
        protocol = 'https' if ssl_enabled else 'http'
        url = f"{protocol}://{host}:{port}"
        
        # 创建远程客户端
        settings = Settings(
            chroma_api_impl="chromadb.api.fastapi.FastAPI",
            chroma_server_host=host,
            chroma_server_http_port=port,
            chroma_server_ssl_enabled=ssl_enabled,
        )
        
        # 如果有认证令牌，添加到headers
        headers = {}
        if auth_token:
            headers['Authorization'] = f'Bearer {auth_token}'
            headers['X-Chroma-Token'] = auth_token
        
        try:
            client = chromadb.HttpClient(
                host=host,
                port=port,
                ssl=ssl_enabled,
                headers=headers
            )
            return client
        except Exception as e:
            raise Exception(f"无法连接到远程 ChromaDB 服务器 {url}: {str(e)}")
    else:
        # 本地连接
        path = config.get('chromadb', 'path', fallback='./data/chromadb')
        return chromadb.PersistentClient(path=path)


@router.get("/config")
async def get_config_info():
    """获取 ChromaDB 配置信息（用于调试）"""
    config, found_path = get_config()
    
    result = {
        "config_file": found_path or "未找到",
        "working_directory": os.getcwd()
    }
    if found_path and config.has_section('chromadb'):
        result.update({
            "chromadb": {
                "connection_type": config.get('chromadb', 'connection_type', fallback='local'),
                "host": config.get('chromadb', 'host', fallback='N/A'),
                "port": config.get('chromadb', 'port', fallback='N/A'),
                "ssl_enabled": config.get('chromadb', 'ssl_enabled', fallback='N/A'),
                "path": config.get('chromadb', 'path', fallback='N/A')
            }
        })
    if found_path and config.has_section('embedding_api'):
        result["embedding_api"] = {
            "endpoint": config.get('embedding_api', 'endpoint', fallback='N/A'),
            "model": config.get('embedding_api', 'model', fallback='N/A'),
            "configured": True
        }
    else:
        result["embedding_api"] = {"configured": False}
    
    if not found_path:
        result["error"] = "配置文件未找到"
    
    return result


@router.get("/collections")
async def get_collections():
    """获取所有集合列表"""
    try:
        client = get_chromadb_client()
        collections = client.list_collections()
        
        collection_info = []
        for col in collections:
            collection_info.append({
                "name": col.name,
                "count": col.count(),
                "metadata": col.metadata
            })
        
        return {"collections": collection_info}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取集合列表失败: {str(e)}")


@router.post("/query")
async def query_collection(request: QueryRequest):
    """查询集合"""
    try:
        client = get_chromadb_client()
        embedding_function = get_embedding_function()
        
        # 获取集合
        try:
            if embedding_function:
                collection = client.get_collection(
                    name=request.collection_name,
                    embedding_function=embedding_function
                )
            else:
                collection = client.get_collection(name=request.collection_name)
        except Exception as e:
            raise HTTPException(status_code=404, detail=f"集合 '{request.collection_name}' 不存在: {str(e)}")
        
        # 执行查询
        try:
            results = collection.query(
                query_texts=[request.query_text],
                n_results=request.n_results
            )
        except Exception as e:
            raise HTTPException(status_code=500, detail=f"查询失败: {str(e)}")
        
        # 格式化结果
        formatted_results = []
        if results['ids'] and len(results['ids']) > 0:
            for i in range(len(results['ids'][0])):
                formatted_results.append({
                    "id": results['ids'][0][i],
                    "document": results['documents'][0][i] if results['documents'] else "",
                    "metadata": results['metadatas'][0][i] if results['metadatas'] else {},
                    "distance": results['distances'][0][i] if results['distances'] else 0.0
                })
        
        return {
            "collection": request.collection_name,
            "query": request.query_text,
            "results": formatted_results,
            "total": len(formatted_results)
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"查询失败: {str(e)}")


@router.get("/collection/{collection_name}/sample")
async def get_collection_sample(collection_name: str, limit: int = 5):
    """获取集合的示例数据"""
    try:
        client = get_chromadb_client()
        
        try:
            collection = client.get_collection(name=collection_name)
        except Exception:
            raise HTTPException(status_code=404, detail=f"集合 '{collection_name}' 不存在")
        
        # 获取前N条数据
        results = collection.get(limit=limit)
        
        samples = []
        if results['ids']:
            for i in range(len(results['ids'])):
                samples.append({
                    "id": results['ids'][i],
                    "document": results['documents'][i] if results['documents'] else "",
                    "metadata": results['metadatas'][i] if results['metadatas'] else {}
                })
        
        return {
            "collection": collection_name,
            "samples": samples,
            "total_count": collection.count()
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取示例数据失败: {str(e)}")
